sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -439,6 +439,39 @@ object StateStoreProvider {
   }
 }
 
+/**
+ * This is an optional trait to be implemented by [[StateStoreProvider]]s that can read fine
+ * grained state data which is replayed from a specific snapshot version. It is used by the
+ * snapshotStartBatchId option in state data source.
+ */
+trait SupportsFineGrainedReplayFromSnapshot {
+  /**
+   * Used by snapshotStartBatchId option when reading state generated by join operation as data
+   * source.
+   * Return an instance of [[StateStore]] representing state data of the given version.
+   * The State Store will be constructed from the snapshot at startVersion, and applying delta files
+   * up to the endVersion. If there is no snapshot file at startVersion, an exception will be
+   * thrown.
+   *
+   * @param snapshotVersion checkpoint version of the snapshot to start with
+   * @param endVersion   checkpoint version to end with
+   */
+  def replayStateFromSnapshot(snapshotVersion: Long, endVersion: Long): StateStore
+
+  /**
+   * Used by snapshotStartBatchId option when reading state generated by all stateful operations
+   * except join as data source.
+   * Return an instance of [[ReadStateStore]] representing state data of the given version.
+   * The State Store will be constructed from the snapshot at startVersion, and applying delta files
+   * up to the endVersion. If there is no snapshot file at startVersion, an exception will be
+   * thrown.
+   *
+   * @param snapshotVersion checkpoint version of the snapshot to start with
+   * @param endVersion   checkpoint version to end with
+   */
+  def replayReadStateFromSnapshot(snapshotVersion: Long, endVersion: Long): ReadStateStore
```

Comment:
Shall we just provide the default implementation to wrap the read-write state store to read-only? We wouldn't need to let state store provider to implement this except the case they can optimize specifically for read-only.

Reply:
You can update the comment like the way why they may want to implement this or they would just leave it as default, instead of describing where will call this method.